ccl/streamingccl/streamingest: TestTenantStreamingProducerJobTimedOut failed

ccl/streamingccl/streamingest.TestTenantStreamingProducerJobTimedOut [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5731396?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5731396?buildTab=artifacts#/) on master @ [687171ac6c2cd9992486bb3b8c9d252ac95ca1cd](https://github.com/cockroachdb/cockroach/commits/687171ac6c2cd9992486bb3b8c9d252ac95ca1cd):


```
=== RUN   TestTenantStreamingProducerJobTimedOut
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/045fecfe9e18c9da6f776a53d446f2c5/logTestTenantStreamingProducerJobTimedOut3649603095
    test_log_scope.go:80: use -show-logs to present logs inline
    panic.go:1047: -- test log scope end --
    panic.go:1047: runtime error: invalid memory address or nil pointer dereference
        goroutine 9578532 [running]:
        runtime/debug.Stack()
        	GOROOT/src/runtime/debug/stack.go:24 +0x65
        github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func1()
        	github.com/cockroachdb/cockroach/pkg/util/leaktest/leaktest.go:110 +0x92
        panic({0x61daae0, 0x9966ae0})
        	GOROOT/src/runtime/panic.go:1047 +0x266
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.TestTenantStreamingProducerJobTimedOut(0xc0153f8820)
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go:330 +0x4af
        testing.tRunner(0xc0153f8820, 0x6b42150)
        	GOROOT/src/testing/testing.go:1259 +0x102
        created by testing.(*T).Run
        	GOROOT/src/testing/testing.go:1306 +0x35a
--- FAIL: TestTenantStreamingProducerJobTimedOut (12.22s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/tenant-streaming
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantStreamingProducerJobTimedOut.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-17591

ccl/streamingccl/streamingest.TestTenantStreamingProducerJobTimedOut [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5754370?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5754370?buildTab=artifacts#/) on master @ [902f937bbcac81e99a3465827e3448b50a55ff87](https://github.com/cockroachdb/cockroach/commits/902f937bbcac81e99a3465827e3448b50a55ff87):


```
=== RUN   TestTenantStreamingProducerJobTimedOut
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/045fecfe9e18c9da6f776a53d446f2c5/logTestTenantStreamingProducerJobTimedOut3634451526
    test_log_scope.go:80: use -show-logs to present logs inline
    panic.go:1047: -- test log scope end --
    panic.go:1047: runtime error: invalid memory address or nil pointer dereference
        goroutine 9908836 [running]:
        runtime/debug.Stack()
        	GOROOT/src/runtime/debug/stack.go:24 +0x65
        github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func1()
        	github.com/cockroachdb/cockroach/pkg/util/leaktest/leaktest.go:110 +0x92
        panic({0x61f0380, 0x998ab10})
        	GOROOT/src/runtime/panic.go:1047 +0x266
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.TestTenantStreamingProducerJobTimedOut(0xc012d4f6c0)
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go:328 +0x4af
        testing.tRunner(0xc012d4f6c0, 0x6b59d50)
        	GOROOT/src/testing/testing.go:1259 +0x102
        created by testing.(*T).Run
        	GOROOT/src/testing/testing.go:1306 +0x35a
--- FAIL: TestTenantStreamingProducerJobTimedOut (10.75s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantStreamingProducerJobTimedOut.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamingest.TestTenantStreamingProducerJobTimedOut [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5765776?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5765776?buildTab=artifacts#/) on master @ [0e816efe5443764591a944edcb385c7b2762523e](https://github.com/cockroachdb/cockroach/commits/0e816efe5443764591a944edcb385c7b2762523e):


```
=== RUN   TestTenantStreamingProducerJobTimedOut
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/045fecfe9e18c9da6f776a53d446f2c5/logTestTenantStreamingProducerJobTimedOut1524118264
    test_log_scope.go:80: use -show-logs to present logs inline
    panic.go:1047: -- test log scope end --
    panic.go:1047: runtime error: invalid memory address or nil pointer dereference
        goroutine 10667389 [running]:
        runtime/debug.Stack()
        	GOROOT/src/runtime/debug/stack.go:24 +0x65
        github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func1()
        	github.com/cockroachdb/cockroach/pkg/util/leaktest/leaktest.go:110 +0x92
        panic({0x61ecca0, 0x9986b10})
        	GOROOT/src/runtime/panic.go:1047 +0x266
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.TestTenantStreamingProducerJobTimedOut(0xc005900340)
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go:328 +0x4af
        testing.tRunner(0xc005900340, 0x6b56270)
        	GOROOT/src/testing/testing.go:1259 +0x102
        created by testing.(*T).Run
        	GOROOT/src/testing/testing.go:1306 +0x35a
--- FAIL: TestTenantStreamingProducerJobTimedOut (10.79s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantStreamingProducerJobTimedOut.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamingest.TestTenantStreamingProducerJobTimedOut [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5771647?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5771647?buildTab=artifacts#/) on master @ [e4cafeb8b1d586d091fb98e3e570650d7eeea294](https://github.com/cockroachdb/cockroach/commits/e4cafeb8b1d586d091fb98e3e570650d7eeea294):


```
=== RUN   TestTenantStreamingProducerJobTimedOut
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/045fecfe9e18c9da6f776a53d446f2c5/logTestTenantStreamingProducerJobTimedOut1961074392
    test_log_scope.go:80: use -show-logs to present logs inline
    panic.go:1047: -- test log scope end --
    panic.go:1047: runtime error: invalid memory address or nil pointer dereference
        goroutine 10437365 [running]:
        runtime/debug.Stack()
        	GOROOT/src/runtime/debug/stack.go:24 +0x65
        github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func1()
        	github.com/cockroachdb/cockroach/pkg/util/leaktest/leaktest.go:110 +0x92
        panic({0x61ee660, 0x998ab10})
        	GOROOT/src/runtime/panic.go:1047 +0x266
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.TestTenantStreamingProducerJobTimedOut(0xc00641e680)
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go:328 +0x4af
        testing.tRunner(0xc00641e680, 0x6b57da8)
        	GOROOT/src/testing/testing.go:1259 +0x102
        created by testing.(*T).Run
        	GOROOT/src/testing/testing.go:1306 +0x35a
--- FAIL: TestTenantStreamingProducerJobTimedOut (11.00s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantStreamingProducerJobTimedOut.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

ccl/streamingccl/streamingest.TestTenantStreamingProducerJobTimedOut [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5777002?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/5777002?buildTab=artifacts#/) on master @ [dc382b1ec54b4599cce1f5f50f243723a4286da6](https://github.com/cockroachdb/cockroach/commits/dc382b1ec54b4599cce1f5f50f243723a4286da6):


```
=== RUN   TestTenantStreamingProducerJobTimedOut
    test_log_scope.go:79: test logs captured to: /artifacts/tmp/_tmp/045fecfe9e18c9da6f776a53d446f2c5/logTestTenantStreamingProducerJobTimedOut1546818604
    test_log_scope.go:80: use -show-logs to present logs inline
    panic.go:1047: -- test log scope end --
    panic.go:1047: runtime error: invalid memory address or nil pointer dereference
        goroutine 10254420 [running]:
        runtime/debug.Stack()
        	GOROOT/src/runtime/debug/stack.go:24 +0x65
        github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func1()
        	github.com/cockroachdb/cockroach/pkg/util/leaktest/leaktest.go:110 +0x92
        panic({0x61ee660, 0x998ab10})
        	GOROOT/src/runtime/panic.go:1047 +0x266
        github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest.TestTenantStreamingProducerJobTimedOut(0xc002a044e0)
        	github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go:328 +0x4af
        testing.tRunner(0xc002a044e0, 0x6b57da8)
        	GOROOT/src/testing/testing.go:1259 +0x102
        created by testing.(*T).Run
        	GOROOT/src/testing/testing.go:1306 +0x35a
--- FAIL: TestTenantStreamingProducerJobTimedOut (11.19s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestTenantStreamingProducerJobTimedOut.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go
+++ b/pkg/ccl/streamingccl/streamingest/stream_ingestion_job.go
@@ -59,7 +59,7 @@ func getStreamIngestionStats(
 	evalCtx *eval.Context, txn *kv.Txn, ingestionJobID jobspb.JobID,
 ) (*streampb.StreamIngestionStats, error) {
 	registry := evalCtx.Planner.ExecutorConfig().(*sql.ExecutorConfig).JobRegistry
-	j, err := registry.LoadJob(evalCtx.Ctx(), ingestionJobID)
+	j, err := registry.LoadJobWithTxn(evalCtx.Ctx(), ingestionJobID, txn)
 	if err != nil {
 		return nil, err
 	}

--- a/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor_planning.go
+++ b/pkg/ccl/streamingccl/streamingest/stream_ingestion_processor_planning.go
@@ -194,7 +194,9 @@ func (s *streamIngestionResultWriter) AddRow(ctx context.Context, row tree.Datum
 		&ingestedHighWatermark); err != nil {
 		return errors.NewAssertionErrorWithWrappedErrf(err, `unmarshalling resolved timestamp`)
 	}
-	return s.registry.UpdateJobWithTxn(ctx, s.jobID, nil /* txn */, false, /* useReadLock */
+	// TODO(casper): currently if this update is without read lock, read may see nil high watermark
+	// when getting a stream ingestion stats. We need to keep investigating why this happens.
+	return s.registry.UpdateJobWithTxn(ctx, s.jobID, nil /* txn */, true, /* useReadLock */
 		func(txn *kv.Txn, md jobs.JobMetadata, ju *jobs.JobUpdater) error {
 			if err := jobs.UpdateHighwaterProgressed(ingestedHighWatermark, md, ju); err != nil {
 				return err

--- a/pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go
+++ b/pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go
@@ -325,6 +325,8 @@ func TestTenantStreamingProducerJobTimedOut(t *testing.T) {
 	c.compareResult("SELECT * FROM d.t2")
 
 	stats := streamIngestionStats(t, c.destSysSQL, ingestionJobID)
+
+	require.NotNil(t, stats.ReplicationLagInfo)
 	require.True(t, srcTime.LessEq(stats.ReplicationLagInfo.MinIngestedTimestamp))
 	require.Equal(t, "", stats.ProducerError)
 

